src/file/aedat4.hpp

```diff
@@ -372,19 +164,276 @@ struct AEDAT4 {
     stream.write(compressed, size);
   }
 
-  static Generator<AER::Event> aedat_to_stream(const std::string filename) {
-    AEDAT4 aedat = AEDAT4(filename);
-    // TODO: Iterate over raw file pointer to save memory
-    for (auto event : aedat.polarity_events) {
-      co_yield event;
+  std::tuple<AER::Event *, size_t> read_events(const int64_t n_events = -1) {
+    int64_t byte_count =
+        (n_events > 0 ? n_events : total_number_of_events) * sizeof(AER::Event);
+    AER::Event *events = (AER::Event *)malloc(byte_count);
+
+    int64_t count = 0;
+    while (get_packet()) {
+      for (; packet_events_read < event_vector->size(); packet_events_read++) {
+        const Event *event = event_vector->Get(packet_events_read);
+        if (n_events > 0 && n_events - count <= 0) {
+          return {events, count};
+        }
+
+        const auto e = AER::Event{
+            static_cast<uint64_t>(event->t()),
+            static_cast<uint16_t>(event->x()),
+            static_cast<uint16_t>(event->y()),
+            static_cast<bool>(event->on()),
+        };
+        events[count] = e;
+        count += 1;
+      }
+    }
+    return {events, count};
+  }
+
+  Generator<AER::Event> stream(const int64_t n_events = -1) {
+    int64_t size = 0, count = 0;
+    static const size_t STREAM_BUFFER_SIZE = 128;
+    do {
+      auto ret = read_events(STREAM_BUFFER_SIZE);
+      auto events = std::get<0>(ret);
+      size = std::get<1>(ret);
+      for (size_t i = 0; i < size; i++) {
+        co_yield events[i];
+        count++;
+      }
+    } while (size == STREAM_BUFFER_SIZE && (n_events < 0 || (count - n_events >= 0)));
```

Comment:
I think the sign in `(count - n_events >= 0)` should be inverted (so `(count - n_events <= 0)`)